### PR TITLE
Allow resizing of file-backed storage, if it's shared.

### PR DESCRIPTION
**Pull request by georgevdd**

### CODE REVIEW

**Comment by soumith** [comment]
I am not knowledgable enough on this. Anyone else want to review this? @andresy @colesbury @dominikgrewe @clementfarabet

**Comment by georgevdd** [comment]
Anyone? Ferris?

**Comment by koraykv** [comment]
This seems good to me, @andresy can you please take a look to make sure there are no corner cases.

File: lib/TH/THAllocator.c [hunk 1]
```diff
@@ -161,8 +161,8 @@ static void *THMapAllocator_alloc(void* ctx_, long size)
     else
       data = MapViewOfFile(hmfile, FILE_MAP_COPY, 0, 0, 0);
 
-    CloseHandle(hfile); 
-    CloseHandle(hmfile); 
+    CloseHandle(hfile);
```
Comment (tkoeppe) [inline lib/TH/THAllocator.c]: Don't the previous two `THError`s leak `hfile`?

File: lib/TH/THAllocator.c [hunk 2]
```diff
@@ -250,6 +245,20 @@ static void THMapAllocator_free(void* ctx_, void* data) {
   THMapAllocatorContext_free(ctx);
```
Comment (tkoeppe) [inline lib/TH/THAllocator.c]: Does the context not get freed if there was an error?

File: lib/TH/THAllocator.c [hunk 3]
```diff
@@ -250,6 +245,20 @@ static void THMapAllocator_free(void* ctx_, void* data) {
...
+
+  THMapAllocatorContext *tmpctx = THMapAllocatorContext_new(
+      ctx->filename, ctx->shared);
+  void *data = THMapAllocator_alloc(tmpctx, size);
+  if (data) {
+    tmpctx->size = ctx->size;
+    ctx->size = size;
+    THMapAllocator_free(tmpctx, ptr);
+  }
+  return data;
```
Comment (tkoeppe) [inline lib/TH/THAllocator.c]: If allocation fails, who cleans up `tmpctx`?

File: lib/TH/THAllocator.c [hunk 4]
```diff
@@ -250,6 +245,20 @@ static void THMapAllocator_free(void* ctx_, void* data) {
   THMapAllocatorContext_free(ctx);
 }
 
+static void *THMapAllocator_realloc(void* ctx_, void* ptr, long size) {
+  THMapAllocatorContext *ctx = ctx_;
+
+  THMapAllocatorContext *tmpctx = THMapAllocatorContext_new(
+      ctx->filename, ctx->shared);
```
Comment (tkoeppe) [inline lib/TH/THAllocator.c]: What if `THMapAllocatorContext_new` fails?

**Comment by soumith** [comment]
@georgevdd there are inline comments. what do you think? if you address them, can you please reopen the PR.

**Pull request closed without merging**